xprof/tally.hpp

```diff
@@ -1,31 +1,820 @@
+  print_tally(std::cout, title, tuple_tally, keys_string);
+  std::cout << std::endl;
+
+  auto aggregated_by_name = aggregate_by_name(m);
+  print_tally(aggregated_by_name, display_name_max_size);
+}
+
+// original_map is map where the key are tuple who correspond to hostname, process, ..., API call
+// name, and the value are TallyCore
+template <typename K, typename T, typename TC, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+void print_extended(std::string title, std::unordered_map<K, TC> m, T &&keys_string, int display_name_max_size) {
+
+  // Now working of the body of the table
+  auto aggregated_nested = aggregate_nested(m);
+  for (auto &[k, aggregated_by_name] : aggregated_nested) {
+    print_named_tuple(std::cout, title, k, keys_string);
+    std::cout << std::endl;
+    print_tally(aggregated_by_name, display_name_max_size);
+  }
+}
+
+//        __  _
+//     | (_  / \ |\ |
+//   \_| __) \_/ | \|
+//
+// https://github.com/nlohmann/json
+//
+template <typename TC, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+void to_json(nlohmann::json &j, const TC &tc) {
+  j = nlohmann::json{{"time", tc.duration}, {"call", tc.count}, {"min", tc.min}, {"max", tc.max}};
+  if (tc.error != 0)
+    j["error"] = tc.error;
+}
+
+template <typename TC, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+void to_json(nlohmann::json &j, const std::vector<std::pair<thapi_function_name, TC>> &aggregated) {
+  for (auto const &[key, val] : aggregated)
+    j[key] = val;
+}
+
+// original_map is map where the key are tuple who correspond to hostname, process, ..., API call
+// name, and the value are TallyCore
+template <typename K, typename TC, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+nlohmann::json json_compact(std::unordered_map<K, TC> &m) {
+  auto aggregated_by_name = aggregate_by_name(m);
+  auto sorted_by_value = sort_by_value(aggregated_by_name);
+  add_footer(sorted_by_value);
+  return {{"data", sorted_by_value}};
+}
+
+template <class... T, class... T2, size_t... I>
+void json_populate(nlohmann::json &j, const std::tuple<T...> &h, const std::tuple<T2...> &s, std::index_sequence<I...>) {
+  ((j[std::get<I>(h)] = std::get<I>(s)), ...);
+}
+
+template <typename K, typename TC, class... T, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+nlohmann::json json_extented(std::unordered_map<K, TC> &m, std::tuple<T...> &&h) {
+  nlohmann::json j;
+  auto aggregated_nested = aggregate_nested(m);
+  for (auto &[k, aggregated_by_name] : aggregated_nested) {
+    auto sorted_by_value = sort_by_value(aggregated_by_name);
+    add_footer(sorted_by_value);
+
+    nlohmann::json j2 = {{"data", sorted_by_value}};
+    constexpr auto seq = std::make_index_sequence<sizeof...(T)>();
+    json_populate(j2, h, k, seq);
+    j.push_back(j2);
+  }
+  return j;
+}
+
+//
+// Metadata
+//
+//
+void print_metadata(std::vector<std::string> metadata) {
```

Comment:
I think this can be in the tally.cpp, should not be in the cpp for sure

Reply:
What should be the criteria for considering moving this?

Reply:
it's function non-templated. The definition/implementation have nothing to do in the `.hpp`.
So you can put the declaration inside the `hpp` and then the implementation in the `cpp`.
But this function is only used on the `cpp`, so no need to export the symbol